// n8n-nodes-confluence/dist/nodes/Confluence/helper/ai-vision.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateImageDescription = generateImageDescription;
async function generateImageDescription(ctx, imageData, filename, config) {
    if (!imageData || imageData.byteLength === 0) {
        throw new Error(`Empty or invalid image data for file ${filename}`);
    }
    const base64Image = Buffer.from(imageData).toString('base64');
    const mimeType = getMimeTypeFromFilename(filename).trim();
    if (!['image/png', 'image/jpeg', 'image/gif', 'image/webp'].includes(mimeType)) {
        throw new Error(`Unsupported image format: ${mimeType} for file ${filename}`);
    }
    switch (config.provider) {
        case 'openai':
            return generateOpenAIDescription(ctx, base64Image, mimeType, config, filename);
        case 'anthropic':
            return generateAnthropicDescription(ctx, base64Image, mimeType, config, filename);
        case 'google':
            return generateGoogleDescription(ctx, base64Image, mimeType, config, filename);
        case 'custom':
            return generateCustomDescription(ctx, base64Image, mimeType, config, filename);
        default:
            throw new Error(`Unsupported AI vision provider: ${config.provider}`);
    }
}
async function generateOpenAIDescription(ctx, base64Image, mimeType, config, filename) {
    var _a, _b, _c, _d, _e, _f;
    const options = {
        method: 'POST',
        url: config.apiUrl || 'https://api.openai.com/v1/chat/completions',
        headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${config.apiKey}`,
        },
        json: true,
        body: {
            model: config.model || 'gpt-4o',
            messages: [
                {
                    role: 'user',
                    content: [
                        {
                            type: 'text',
                            text: 'Beschreibe dieses Bild auf Deutsch in 1-2 prägnanten Sätzen. Fokussiere auf die wichtigsten visuellen Elemente, die für das Verständnis des Inhalts relevant sind. Beginne nicht mit "Das Bild zeigt" oder ähnlichen Phrasen.',
                        },
                        {
                            type: 'image_url',
                            image_url: {
                                url: `data:${mimeType};base64,${base64Image}`,
                            },
                        },
                    ],
                },
            ],
            max_tokens: config.maxTokens || 150,
            temperature: config.temperature || 0.3,
        },
    };
    try {
        const response = await ctx.helpers.request(options);
        return (((_d = (_c = (_b = (_a = response.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.message) === null || _c === void 0 ? void 0 : _c.content) === null || _d === void 0 ? void 0 : _d.trim()) || 'Bild konnte nicht beschrieben werden.');
    }
    catch (error) {
        console.error('OpenAI Vision API Error:', {
            error: error.message,
            status: (_e = error.response) === null || _e === void 0 ? void 0 : _e.status,
            data: (_f = error.response) === null || _f === void 0 ? void 0 : _f.data,
            filename: filename || 'unknown',
            mimeType: mimeType,
        });
        throw error;
    }
}
async function generateAnthropicDescription(ctx, base64Image, mimeType, config, filename) {
    var _a, _b, _c;
    const options = {
        method: 'POST',
        url: config.apiUrl || 'https://api.anthropic.com/v1/messages',
        headers: {
            'Content-Type': 'application/json',
            'x-api-key': config.apiKey,
            'anthropic-version': '2023-06-01',
        },
        json: true,
        body: {
            model: config.model || 'claude-3-sonnet-20240229',
            max_tokens: config.maxTokens || 150,
            temperature: config.temperature || 0.3,
            messages: [
                {
                    role: 'user',
                    content: [
                        {
                            type: 'text',
                            text: 'Beschreibe dieses Bild auf Deutsch in 1-2 prägnanten Sätzen. Fokussiere auf die wichtigsten visuellen Elemente, die für das Verständnis des Inhalts relevant sind. Beginne nicht mit "Das Bild zeigt" oder ähnlichen Phrasen.',
                        },
                        {
                            type: 'image',
                            source: {
                                type: 'base64',
                                media_type: mimeType,
                                data: base64Image,
                            },
                        },
                    ],
                },
            ],
        },
    };
    const response = await ctx.helpers.request(options);
    return ((_c = (_b = (_a = response.content) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.text) === null || _c === void 0 ? void 0 : _c.trim()) || 'Bild konnte nicht beschrieben werden.';
}
async function generateGoogleDescription(ctx, base64Image, mimeType, config, filename) {
    var _a, _b, _c, _d, _e, _f;
    const options = {
        method: 'POST',
        url: config.apiUrl ||
            `https://generativelanguage.googleapis.com/v1beta/models/${config.model || 'gemini-pro-vision'}:generateContent?key=${config.apiKey}`,
        headers: {
            'Content-Type': 'application/json',
        },
        json: true,
        body: {
            contents: [
                {
                    parts: [
                        {
                            text: 'Beschreibe dieses Bild auf Deutsch in 1-2 prägnanten Sätzen. Fokussiere auf die wichtigsten visuellen Elemente, die für das Verständnis des Inhalts relevant sind. Beginne nicht mit "Das Bild zeigt" oder ähnlichen Phrasen.',
                        },
                        {
                            inline_data: {
                                mime_type: mimeType,
                                data: base64Image,
                            },
                        },
                    ],
                },
            ],
            generationConfig: {
                maxOutputTokens: config.maxTokens || 150,
                temperature: config.temperature || 0.3,
            },
        },
    };
    const response = await ctx.helpers.request(options);
    return (((_f = (_e = (_d = (_c = (_b = (_a = response.candidates) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.content) === null || _c === void 0 ? void 0 : _c.parts) === null || _d === void 0 ? void 0 : _d[0]) === null || _e === void 0 ? void 0 : _e.text) === null || _f === void 0 ? void 0 : _f.trim()) ||
        'Bild konnte nicht beschrieben werden.');
}
async function generateCustomDescription(ctx, base64Image, mimeType, config, filename) {
    var _a, _b;
    const options = {
        method: 'POST',
        url: config.apiUrl,
        headers: {
            'Content-Type': 'application/json',
            ...(config.apiKey && { Authorization: `Bearer ${config.apiKey}` }),
        },
        json: true,
        body: {
            image: base64Image,
            mimeType,
            prompt: 'Beschreibe dieses Bild auf Deutsch in 1-2 prägnanten Sätzen. Fokussiere auf die wichtigsten visuellen Elemente, die für das Verständnis des Inhalts relevant sind. Beginne nicht mit "Das Bild zeigt" oder ähnlichen Phrasen.',
            maxTokens: config.maxTokens || 150,
            temperature: config.temperature || 0.3,
        },
    };
    const response = await ctx.helpers.request(options);
    return (((_a = response.description) === null || _a === void 0 ? void 0 : _a.trim()) || ((_b = response.text) === null || _b === void 0 ? void 0 : _b.trim()) || 'Bild konnte nicht beschrieben werden.');
}
function getMimeTypeFromFilename(filename) {
    const ext = filename.toLowerCase().split('.').pop();
    switch (ext) {
        case 'png':
            return 'image/png';
        case 'jpg':
        case 'jpeg':
            return 'image/jpeg';
        case 'gif':
            return 'image/gif';
        case 'webp':
            return 'image/webp';
        case 'bmp':
            return 'image/bmp';
        case 'svg':
            return 'image/svg+xml';
        default:
            return 'image/png';
    }
}
