// n8n-nodes-confluence/dist/nodes/Confluence/helper/images.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractImageReferences = extractImageReferences;
exports.fetchPageAttachments = fetchPageAttachments;
exports.matchImagesToAttachments = matchImagesToAttachments;
exports.downloadImage = downloadImage;
exports.replaceImageTagsWithDescriptions = replaceImageTagsWithDescriptions;
exports.isImageAttachment = isImageAttachment;
const models_1 = require("../models");
const validation_1 = require("./validation");
function extractImageReferences(html) {
    if (!html)
        return [];
    const images = [];
    const imageTagRegex = /<ac:image[^>]*>.*?<\/ac:image>/gis;
    const imageMatches = html.match(imageTagRegex) || [];
    for (const imageTag of imageMatches) {
        const filenameMatch = imageTag.match(/ri:filename="([^"]+)"/i);
        if (filenameMatch) {
            images.push({
                filename: filenameMatch[1],
                originalTag: imageTag,
            });
        }
    }
    return images;
}
async function fetchPageAttachments(ctx, baseUrl, pageId, itemIndex) {
    const options = {
        headers: { Accept: 'application/json' },
        method: 'GET',
        url: `${baseUrl}/rest/api/content/${encodeURIComponent(pageId)}/child/attachment`,
        json: true,
    };
    const raw = await ctx.helpers.requestWithAuthentication.call(ctx, 'confluenceCredentialsApi', options);
    const parsed = (0, validation_1.parseOrThrow)(ctx, models_1.attachmentPageSchema, raw, `Failed to fetch attachments for page ${pageId}.`, itemIndex);
    return parsed.results;
}
function matchImagesToAttachments(imageRefs, attachments) {
    return imageRefs.map((img) => {
        const attachment = attachments.find((att) => att.title === img.filename);
        return {
            ...img,
            attachmentId: attachment === null || attachment === void 0 ? void 0 : attachment.id,
        };
    });
}
async function downloadImage(ctx, baseUrl, downloadPath) {
    const options = {
        headers: { Accept: 'image/*' },
        method: 'GET',
        url: `${baseUrl}${downloadPath}`,
        json: false,
        encoding: null,
    };
    try {
        const imageData = await ctx.helpers.requestWithAuthentication.call(ctx, 'confluenceCredentialsApi', options);
        if (!Buffer.isBuffer(imageData)) {
            throw new Error(`Expected Buffer but received ${typeof imageData}`);
        }
        if (imageData.length === 0) {
            throw new Error(`Received empty image data from ${options.url}`);
        }
        return imageData.buffer.slice(imageData.byteOffset, imageData.byteOffset + imageData.byteLength);
    }
    catch (error) {
        throw new Error(`Failed to download image from ${options.url}: ${error}`);
    }
}
function replaceImageTagsWithDescriptions(html, imageRefs) {
    let processedHtml = html;
    for (const img of imageRefs) {
        if (img.description) {
            processedHtml = processedHtml.replace(img.originalTag, `<span class="image-description">[Bild: ${img.description}]</span>`);
        }
    }
    return processedHtml;
}
function isImageAttachment(attachment) {
    var _a, _b;
    const mediaType = ((_a = attachment.metadata) === null || _a === void 0 ? void 0 : _a.mediaType) || ((_b = attachment.extensions) === null || _b === void 0 ? void 0 : _b.mediaType) || '';
    return mediaType.startsWith('image/');
}
