// n8n-nodes-confluence/dist/credentials/ConfluenceCredentialsApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfluenceCredentialsApi = void 0;
class ConfluenceCredentialsApi {
    constructor() {
        this.name = 'confluenceCredentialsApi';
        this.displayName = 'Confluence Credentials API';
        this.documentationUrl = 'https://your-docs-url';
        this.properties = [
            {
                displayName: 'APIM Subscription Key',
                name: 'apimSubscriptionKey',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'User Name',
                name: 'username',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                auth: {
                    username: '={{ $credentials.username }}',
                    password: '={{ $credentials.password }}',
                },
                headers: {
                    'Ocp-Apim-Subscription-Key': '={{ $credentials.apimSubscriptionKey }}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://example.com/',
                url: '',
            },
        };
    }
}
exports.ConfluenceCredentialsApi = ConfluenceCredentialsApi;
