// n8n-nodes-confluence/dist/nodes/Confluence/Confluence.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Confluence = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const Confluence_node_description_1 = require("./Confluence.node.description");
const models_1 = require("./models");
const transformations_1 = require("./helper/transformations");
const validation_1 = require("./helper/validation");
const images_1 = require("./helper/images");
const ai_vision_1 = require("./helper/ai-vision");
class Confluence {
    constructor() {
        this.description = {
            displayName: 'Confluence',
            name: 'confluence',
            icon: 'file:Confluence.svg',
            group: ['input'],
            version: 1,
            description: 'Confluence datacenter node',
            defaults: {
                name: 'Confluence',
            },
            inputs: ["main"],
            outputs: ["main"],
            usableAsTool: true,
            credentials: [
                {
                    name: 'confluenceCredentialsApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Base URL',
                    name: 'baseUrl',
                    type: 'string',
                    default: '',
                    required: true,
                    description: 'The base URL of your Confluence instance',
                },
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    required: true,
                    options: [
                        {
                            name: 'Space',
                            value: 'space',
                        },
                    ],
                    default: 'space',
                },
                ...Confluence_node_description_1.spaceOperations,
                ...Confluence_node_description_1.spaceFields,
            ],
        };
    }
    async execute() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z;
        const items = this.getInputData();
        const baseUrl = this.getNodeParameter('baseUrl', 0);
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        const returnData = [];
        if (resource === 'space') {
            if (operation === 'getSpaces') {
                const limit = 100;
                let start = 0;
                while (true) {
                    const options = {
                        headers: { Accept: 'application/json' },
                        method: 'GET',
                        url: `${baseUrl}/rest/api/space`,
                        json: true,
                        qs: { limit, start },
                    };
                    const pageRaw = await this.helpers.requestWithAuthentication.call(this, 'confluenceCredentialsApi', options);
                    const parsed = (0, validation_1.parseOrThrow)(this, models_1.spacePageSchema, pageRaw, 'Invalid response from Confluence API while fetching spaces.', 0);
                    const spaces = parsed.results;
                    for (const s of spaces) {
                        returnData.push({
                            json: {
                                id: s.id,
                                key: s.key,
                                name: s.name,
                                type: s.type,
                                _links: {
                                    webui: (_a = s._links) === null || _a === void 0 ? void 0 : _a.webui,
                                    self: (_b = s._links) === null || _b === void 0 ? void 0 : _b.self,
                                },
                            },
                        });
                    }
                    const pageCount = spaces.length;
                    if (pageCount < limit)
                        break;
                    start += pageCount;
                }
            }
            if (operation === 'getSpaceContent') {
                for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
                    try {
                        const spaceKey = this.getNodeParameter('spaceKey', itemIndex, '');
                        const spaceName = this.getNodeParameter('spaceName', itemIndex, '');
                        const processImages = this.getNodeParameter('processImages', itemIndex, false);
                        let aiConfig = null;
                        if (processImages) {
                            const apiKey = this.getNodeParameter('aiApiKey', itemIndex, '');
                            if (!apiKey) {
                                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'AI Vision API key is required when image processing is enabled', { itemIndex });
                            }
                            aiConfig = {
                                provider: this.getNodeParameter('aiProvider', itemIndex, 'openai'),
                                apiKey,
                                model: this.getNodeParameter('aiModel', itemIndex, ''),
                                apiUrl: this.getNodeParameter('aiApiUrl', itemIndex, ''),
                                maxTokens: 150,
                                temperature: 0.3,
                            };
                        }
                        const item = items[itemIndex];
                        (_d = (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info) === null || _d === void 0 ? void 0 : _d.call(_c, 'Confluence: getSpaceContent', {
                            node: this.getNode().name,
                        });
                        (_f = (_e = this.logger) === null || _e === void 0 ? void 0 : _e.debug) === null || _f === void 0 ? void 0 : _f.call(_e, `Confluence: getSpaceContent for space ${spaceKey}`);
                        const limit = 100;
                        let start = 0;
                        const pages = [];
                        while (true) {
                            const options = {
                                headers: { Accept: 'application/json' },
                                method: 'GET',
                                url: `${baseUrl}/rest/api/space/${encodeURIComponent(spaceKey)}/content`,
                                json: true,
                                qs: { expand: 'body.storage', limit, start },
                            };
                            (_h = (_g = this.logger) === null || _g === void 0 ? void 0 : _g.debug) === null || _h === void 0 ? void 0 : _h.call(_g, `Confluence: getSpaceContent for space ${spaceKey}`);
                            const raw = await this.helpers.requestWithAuthentication.call(this, 'confluenceCredentialsApi', options);
                            const parsed = (0, validation_1.parseOrThrow)(this, models_1.spaceContentEnvelopeSchema, raw, 'Invalid response from Confluence API while fetching space content.', itemIndex);
                            const parsedPages = parsed.page.results;
                            (_k = (_j = this.logger) === null || _j === void 0 ? void 0 : _j.debug) === null || _k === void 0 ? void 0 : _k.call(_j, `Pages found: ${parsedPages.length}`);
                            for (const p of parsedPages) {
                                let processedBody = p.body.storage.value || '';
                                if (processImages && aiConfig) {
                                    try {
                                        (_m = (_l = this.logger) === null || _l === void 0 ? void 0 : _l.debug) === null || _m === void 0 ? void 0 : _m.call(_l, `Processing images for page ${p.id}`);
                                        const imageRefs = (0, images_1.extractImageReferences)(processedBody);
                                        if (imageRefs.length > 0) {
                                            (_p = (_o = this.logger) === null || _o === void 0 ? void 0 : _o.debug) === null || _p === void 0 ? void 0 : _p.call(_o, `Found ${imageRefs.length} images in page ${p.id}`);
                                            const attachments = await (0, images_1.fetchPageAttachments)(this, baseUrl, p.id, itemIndex);
                                            const imageAttachments = attachments.filter(images_1.isImageAttachment);
                                            const matchedImages = (0, images_1.matchImagesToAttachments)(imageRefs, imageAttachments);
                                            for (const img of matchedImages) {
                                                if (img.attachmentId) {
                                                    const attachment = imageAttachments.find((a) => a.id === img.attachmentId);
                                                    if (attachment) {
                                                        try {
                                                            const imageData = await (0, images_1.downloadImage)(this, baseUrl, attachment._links.download || '');
                                                            img.description = await (0, ai_vision_1.generateImageDescription)(this, imageData, img.filename, aiConfig);
                                                        }
                                                        catch (imgError) {
                                                            (_r = (_q = this.logger) === null || _q === void 0 ? void 0 : _q.warn) === null || _r === void 0 ? void 0 : _r.call(_q, `Failed to process image ${img.filename}:`, imgError);
                                                            img.description = `Bild: ${img.filename} (Beschreibung nicht verfügbar)`;
                                                        }
                                                    }
                                                }
                                                else {
                                                    img.description = `Bild: ${img.filename} (Anhang nicht gefunden)`;
                                                }
                                            }
                                            processedBody = (0, images_1.replaceImageTagsWithDescriptions)(processedBody, matchedImages);
                                        }
                                    }
                                    catch (imgProcessError) {
                                        (_t = (_s = this.logger) === null || _s === void 0 ? void 0 : _s.warn) === null || _t === void 0 ? void 0 : _t.call(_s, `Failed to process images for page ${p.id}:`, imgProcessError);
                                    }
                                }
                                const plainText = (0, transformations_1.buildPagePlainText)(spaceName || '', spaceKey || '', p.title || '', processedBody);
                                pages.push({
                                    spaceKey,
                                    spaceName,
                                    id: p.id,
                                    title: p.title,
                                    body: p.body.storage.value,
                                    plainText,
                                    webuiLink: p._links.webui,
                                });
                            }
                            const count = parsedPages.length;
                            (_v = (_u = this.logger) === null || _u === void 0 ? void 0 : _u.debug) === null || _v === void 0 ? void 0 : _v.call(_u, `Pagination check: count=${count}, limit=${limit}, should break=${count === 0 || count < limit}`);
                            if (count === 0 || count < limit) {
                                (_x = (_w = this.logger) === null || _w === void 0 ? void 0 : _w.debug) === null || _x === void 0 ? void 0 : _x.call(_w, `Breaking pagination loop: processed ${pages.length} total pages`);
                                break;
                            }
                            start += count;
                            (_z = (_y = this.logger) === null || _y === void 0 ? void 0 : _y.debug) === null || _z === void 0 ? void 0 : _z.call(_y, `Continuing pagination: new start=${start}`);
                        }
                        item.json.content = pages;
                    }
                    catch (error) {
                        if (this.continueOnFail()) {
                            items.push({
                                json: this.getInputData(itemIndex)[0].json,
                                error,
                                pairedItem: itemIndex,
                            });
                        }
                        else {
                            if (error.context) {
                                error.context.itemIndex = itemIndex;
                                throw error;
                            }
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), error, {
                                itemIndex,
                            });
                        }
                    }
                }
                return [items];
            }
        }
        return [returnData];
    }
}
exports.Confluence = Confluence;
