// n8n-nodes-confluence/dist/nodes/Confluence/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.imageReferenceSchema = exports.attachmentPageSchema = exports.attachmentSchema = exports.parsedPageSchema = exports.spaceContentEnvelopeSchema = exports.contentPageSchema = exports.linkSchema = exports.contentBodySchema = exports.storageBodySchema = exports.spacePageSchema = exports.spaceSchema = void 0;
const zod_1 = require("zod");
exports.spaceSchema = zod_1.z.object({
    id: zod_1.z.number(),
    key: zod_1.z.string(),
    name: zod_1.z.string(),
    type: zod_1.z.string(),
    _links: zod_1.z
        .object({
        webui: zod_1.z.string().optional(),
        self: zod_1.z.string().optional(),
    })
        .optional(),
});
exports.spacePageSchema = zod_1.z.object({
    results: zod_1.z.array(exports.spaceSchema),
    start: zod_1.z.number().optional(),
    limit: zod_1.z.number().optional(),
    size: zod_1.z.number().optional(),
    _links: zod_1.z.unknown().optional(),
});
exports.storageBodySchema = zod_1.z.object({
    value: zod_1.z.string(),
    representation: zod_1.z.string().optional(),
    _expandable: zod_1.z.unknown().optional(),
});
exports.contentBodySchema = zod_1.z.object({
    storage: exports.storageBodySchema,
    _expandable: zod_1.z.unknown().optional(),
});
exports.linkSchema = zod_1.z.object({
    webui: zod_1.z.string(),
});
exports.contentPageSchema = zod_1.z.object({
    id: zod_1.z.string(),
    type: zod_1.z.literal('page').or(zod_1.z.string()),
    status: zod_1.z.string().optional(),
    title: zod_1.z.string(),
    body: exports.contentBodySchema,
    _links: exports.linkSchema,
    _expandable: zod_1.z.unknown().optional(),
});
exports.spaceContentEnvelopeSchema = zod_1.z.object({
    page: zod_1.z.object({
        results: zod_1.z.array(exports.contentPageSchema),
        start: zod_1.z.number().optional(),
        limit: zod_1.z.number().optional(),
        size: zod_1.z.number().optional(),
        _links: zod_1.z
            .object({ self: zod_1.z.string().optional(), next: zod_1.z.string().optional() })
            .partial()
            .optional(),
    }),
    blogpost: zod_1.z
        .object({
        results: zod_1.z.array(zod_1.z.unknown()),
        start: zod_1.z.number().optional(),
        limit: zod_1.z.number().optional(),
        size: zod_1.z.number().optional(),
        _links: zod_1.z.unknown().optional(),
    })
        .optional(),
    _links: zod_1.z.unknown().optional(),
});
exports.parsedPageSchema = zod_1.z.object({
    spaceKey: zod_1.z.string(),
    spaceName: zod_1.z.string().optional().default(''),
    id: zod_1.z.string(),
    title: zod_1.z.string(),
    body: zod_1.z.string(),
    plainText: zod_1.z.string(),
    webuiLink: zod_1.z.string().optional().default(''),
});
exports.attachmentSchema = zod_1.z.object({
    id: zod_1.z.string(),
    type: zod_1.z.literal('attachment'),
    status: zod_1.z.string(),
    title: zod_1.z.string(),
    metadata: zod_1.z
        .object({
        mediaType: zod_1.z.string(),
        _expandable: zod_1.z.unknown().optional(),
    })
        .optional(),
    extensions: zod_1.z
        .object({
        mediaType: zod_1.z.string(),
        fileSize: zod_1.z.number().optional(),
    })
        .optional(),
    _links: zod_1.z.object({
        download: zod_1.z.string(),
        thumbnail: zod_1.z.string().optional(),
        self: zod_1.z.string().optional(),
        webui: zod_1.z.string().optional(),
    }),
    _expandable: zod_1.z.unknown().optional(),
});
exports.attachmentPageSchema = zod_1.z.object({
    results: zod_1.z.array(exports.attachmentSchema),
    start: zod_1.z.number().optional(),
    limit: zod_1.z.number().optional(),
    size: zod_1.z.number().optional(),
    _links: zod_1.z.unknown().optional(),
});
exports.imageReferenceSchema = zod_1.z.object({
    filename: zod_1.z.string(),
    originalTag: zod_1.z.string(),
    attachmentId: zod_1.z.string().optional(),
    description: zod_1.z.string().optional(),
});
